### ISSUE TITLE
[Feat] Updates on the learning process, model, and analysis

### ISSUE THREAD

**Issue by tixianw** [hanson-hschang/Signal-System#37]
- [ ] Process
  - [x] 'eval_nbatch' determine the number of batches to use when calculating the loss (validation). This allows to use only partial validation dataset to evaluate the model and reduce computation time.
  - [ ] Evaluation metrics
  - [x] `torch.compile`
  - [ ] activation checkpointing
- [ ] Model
  - [x] The option of having different input output layer weights
  - [x] Different options of transition matrix: iid, identity
  - [x] Weights of transition matrix be the same as initial distribution
  - [x] Fix weights to unlearnable: `requires_grad=False`
  - [x] Different probability generation methods
  - [ ] <s>Bounds on parameters</s> wont need it now
- [x] Analysis
  - [x] Check the weights distribution
  - [x] Check initial_state
